033-OSSDatabaseMigration/Coach/03-offline-migration.md

```diff
@@ -0,0 +1,119 @@
+
+```bash
+
+kubectl -n postgresql exec deploy/postgres -it -- bash
+root@postgres-64786b846-khk28:/#  psql -h pgtarget.postgres.database.azure.com -p 5432 -U serveradmin@pgtarget -d postgres
+
+```
+
+Before migrating the data, they need to create an empty database and create the application user. Connect to the database container first and from there connect to Azure DB.
+Alternately connect to Azure DB using  Azure Data studio or Pgadmin tool. The example below connects to postgres on-premise database container. You can run it from azure 
+cloud shell also. 
+
+```bash
+
+kubectl -n postgresql exec deploy/postgres -it -- bash
+psql -h pgtarget.postgres.database.azure.com -p 5432 -U serveradmin@pgtarget -d postgres
+create database wth ;
+
+```
+
+```bash
+
+kubectl -n mysql exec deploy/mysql -it -- bash
+mysql -h mytarget2.mysql.database.azure.com -P 3306 -u serveradmin@mytarget2 -p
+create database wth ;
+
+```
+
+Create the pizzeria application database user and the database wth
+
+```sql
+
+CREATE ROLE CONTOSOAPP WITH LOGIN NOSUPERUSER INHERIT CREATEDB CREATEROLE NOREPLICATION PASSWORD 'OCPHack8';
+create database wth ;
+
+```
+* Postgres command to do offline export to exportdir directory and import offline to Azure Postgres. First bash into the postgres container and then two commands
+
+```bash
+
+ pg_dump -C -Fd  wth -j 4 -f exportdir -U contosoapp
+ pg_restore -h pgtarget.postgres.database.azure.com -p 5432 -U contosoapp@pgtarget -d wth -Fd exportdir
+
+```
+
+* For MySQL the MySQL database script file may contain references to @@SESSION and @@GLOBAL that will need to be removed prior to importing.
+
+Privileges required to run Mysql export on the source database - connect as root
+
+```sql
+grant ALL PRIVILEGES ON wth.* TO contosoapp ;
+grant process, select on *.*  to contosoapp ;
+
+-- check privileges already granted
+
+show grants for contosoapp ;
+
+```
+Grants for contosoapp should report
+
+
+GRANT SELECT, PROCESS ON *.* TO 'contosoapp'@'%'
+
+GRANT ALL PRIVILEGES ON `wth`.* TO 'contosoapp'@'%'
+
+It is possible to use mysql workbench tool to run the export with proper settings.  The mysql workbench version ( 8.0.23 as of Jan 2021 ) being different from mysql version
+5.7 is not a factor for this challenge. The mysql export runs series of exports for each table like this. If you do not want to see the warnings about --set-gtid-purged, use
+the flag  --set-gtid-purged
+
+Running: mysqldump.exe --defaults-file="C:\Users\susengu\AppData\Local\Temp\tmp_akgbble.cnf"  --host=<container ip> --port=3306 --default-character-set=utf8 --user=contosoapp --protocol=tcp --no-data --skip-triggers --skip-column-statistics "wth" "bakestyle" --set-gtid-purged=off
```

Comment:
Remove the word "Running:" and make this a bash code block.
Above the code block put something like "Now run this:"

Reply:
Fixed